### PR TITLE
fix: show evaluated value for action selector fields

### LINKED ISSUE #12736: [Bug]-[3000]:Evaluated value pane does not show the right value for bindings
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

The action creator doesn't show the proper evaluated value for the binding `{{Text1.text}}`, `{{JsObject.myVar1}}` etc....
On execution, the value is processed properly.

image

### Steps To Reproduce

[LOOM DEMO](https://www.loom.com/share/f6d5752f27cb4ce795848444be65de7d)

### Public Sample App

_No response_

### Version

Cloud - Appsmith v1.6.19-SNAPSHOT

### PR DESCRIPTION
**Pull request by ravikp7**
## Description
The evaluated values for text fields in action selector were not shown. This PR fixes the issue.

#### PR fixes following issue(s)
Fixes #12736 

#### Type of change
- Bug fix (non-breaking change which fixes an issue)

#### How Has This Been Tested?
- [x] Cypress
>
>
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Test-plan-implementation#speedbreaker-features-to-consider-for-every-change) have been covered
- [x] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans/_edit#areas-of-interest)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [x] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [x] Cypress test cases have been added and approved by SDET/manual QA
- [x] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/ActionExecution/FieldEvaluation_spec.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import * as _ from "../../../../support/Objects/ObjectsCore";
+
+const { agHelper, apiPage, entityExplorer, propPane } = _;
+
+describe("Field value evaluation", () => {
+  before(() => {
+    cy.fixture("buttondsl").then((val: any) => {
+      agHelper.AddDsl(val);
+    });
+    apiPage.CreateApi("Api1", "GET");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/ActionExecution/FieldEvaluation_spec.ts:10]: Why is API needed here?

Comment (ravikp7) [reply app/client/cypress/integration/Regression_TestSuite/ClientSideTests/ActionExecution/FieldEvaluation_spec.ts]: I was using it before, forgot to remove. Removed now

**Review by Aishwarya-U-R — COMMENTED** [review]
Cypress looks good.

File: app/client/src/ce/constants/ReduxActionConstants.tsx [hunk 2]
```diff
@@ -815,6 +815,10 @@ export const ReduxActionTypes = {
   LOAD_FILE_PICKER_ACTION: "LOAD_FILE_PICKER_ACTION",
   TOGGLE_AI_WINDOW: "TOGGLE_AI_WINDOW",
   UPDATE_DATASOURCE_AUTH_STATE: "UPDATE_DATASOURCE_AUTH_STATE",
+  EVALUATE_ACTION_SELECTOR_FIELD: "EVALUATE_ACTION_SELECTOR_FIELD",
+  SET_EVALUATED_ACTION_SELECTOR_FIELD: "SET_EVALUATED_ACTION_SELECTOR_FIELD",
+  CLEAR_EVALUATED_ACTION_SELECTOR_FIELD:
+    "CLEAR_EVALUATED_ACTION_SELECTOR_FIELD",
```
Comment (rimildeyjsr) [inline app/client/src/ce/constants/ReduxActionConstants.tsx:821]: this redux action types object has been too large, can we make a separate object for action selector and put these actions there

File: app/client/src/components/editorComponents/ActionCreator/viewComponents/TextView/index.tsx [hunk 3]
```diff
@@ -5,9 +5,52 @@ import {
...
+
+  const dispatch = useDispatch();
+
+  const codeWithoutMoustache = getCodeFromMoustache(textValue);
+
+  useEffect(() => {
+    // If the code contains a function or arrow function, don't evaluate it
+    // Because, the evaluations are done in the worker and we can't post functions to the worker
+    if (
+      ["function", "=>"].some((keyword) => codeWithoutMoustache.match(keyword))
```
Comment (rimildeyjsr) [inline app/client/src/components/editorComponents/ActionCreator/viewComponents/TextView/index.tsx:30]: can we use a more generic way to figure out if what we have is a function? Lets not hard code these things

File: app/client/src/ce/constants/ReduxActionConstants.tsx (hunk 2 above)
Comment (ravikp7) [reply app/client/src/ce/constants/ReduxActionConstants.tsx]: Done

File: app/shared/ast/src/index.test.ts [hunk 4]
```diff
@@ -608,3 +608,37 @@ describe("parseJSObjectWithAST", () => {
     expect(parsedObject).toStrictEqual(expectedParsedObject);
   });
 });
+
+describe("isFunctionPresent", () => {
```
Comment (rimildeyjsr) [inline app/shared/ast/src/index.test.ts:612]: let's add IIFEs, arrow function shorthand syntax, higher-order functions, functions with promises and functions with parameters as well, the more extensive these cases are, the more robust are platform will be.

File: app/shared/ast/src/index.ts [hunk 5]
```diff
@@ -749,3 +749,33 @@ const jsObjectToCode = (script: string) => {
...
+) => {
+  try {
+    const sanitizedScript = sanitizeScript(script, evaluationVersion);
+    const ast = getAST(sanitizedScript, {
+      locations: true,
+      ranges: true,
+    });
+
+    let isFunction = false;
+    simple(ast, {
```
Comment (rimildeyjsr) [inline app/shared/ast/src/index.ts:765]: This looks good, will it work well with higher-order functions as well? or functions with functions within them?
Just test once before passing to QA

**Review by rimildeyjsr — APPROVED** [review]
Approving the PR since it looks good - @ravikp7 do add more test cases before passing to QA!

File: app/shared/ast/src/index.test.ts (hunk 4 above)
Comment (ravikp7) [reply app/shared/ast/src/index.test.ts:612]: added test cases

File: app/shared/ast/src/index.ts (hunk 5 above)
Comment (ravikp7) [reply app/shared/ast/src/index.ts:765]: added

**Comment by ramsaptami** [comment]
lgtm
cc: @ravikp7

**Review by rimildeyjsr — APPROVED** [review]

**Pull request merged**